### ISSUE TITLE
auth.json sharing with upstream TS pi is broken twice over: incompatible lock protocol (leftover flock file → permanent ENOTDIR) and incompatible OAuth credential serialization (o_auth/access_token vs oauth/access)

### ISSUE THREAD

**Issue by takumi3488** [Dicklesworthstone/pi_agent_rust#120]
## Summary

pi_agent_rust and upstream TS pi (`@earendil-works/pi-coding-agent`) share `~/.pi/agent/auth.json` (README: "Credentials are stored in `~/.pi/agent/auth.json` with file-locked access to prevent corruption from concurrent instances"), but running them side by side breaks in **two independent ways**:

1. the two implementations use **mutually incompatible lock protocols** on the same lock path `~/.pi/agent/auth.json.lock`, and
2. they use **mutually incompatible serializations for OAuth credentials**, so neither implementation can read an OAuth login performed by the other.

Both surface as misleading auth errors ("Authentication failed", "No API key for provider: ...") while perfectly valid credentials sit in `auth.json`.

Verified against pi_agent_rust `0.1.20` and `@earendil-works/pi-coding-agent` `0.80.2` on macOS.

## Incompatibility 1: lock protocol

| | pi_agent_rust | TS pi (`proper-lockfile@4.1.2`) |
|---|---|---|
| Lock entity | `flock(2)` (via `fs4`) on a persistent **empty regular file** | **directory** created with `mkdir(2)`; *existence = held* |
| Release | kernel releases flock; **file is never deleted** (`src/auth.rs` `LockedFile::drop` only unlocks) | `rmdir` on release; mtime-based stale reclaim via `rmdir` |

Once pi_agent_rust runs even once, it leaves `auth.json.lock` behind as a 0-byte regular file. From that point, every TS pi credential access fails:

1. **Fresh mtime** (rust pi recently touched it): `mkdir` → `EEXIST`, not stale → `ELOCKED: Lock file is already being held`. TS pi's sync read path retries only 10×20ms, so it gives up in ~200ms.
2. **Stale mtime**: proper-lockfile tries to reclaim with `rmdir(2)` → **`ENOTDIR: not a directory, rmdir '.../auth.json.lock'`** because the lock is a regular file. **The lock can never be reclaimed — auth loading is permanently poisoned until the file is deleted by hand** (and it reappears on the next rust pi run).
3. Reverse direction: while TS pi holds the lock (a directory), rust pi's `open(auth.json.lock, O_CREAT)` fails with `EISDIR` → `auth lock file: ...`.

TS pi swallows the lock error inside `AuthStorage.reload()` (in-memory credentials stay empty), so the user-visible symptom is a misleading auth error, e.g.:

```
Authentication failed for "anthropic". Credentials may have expired or network
is unavailable. Run '/login anthropic' to re-authenticate.
```

This exact `ENOTDIR` symptom was independently reported by a user in a comment on upstream issue earendil-works/pi#1871 ("It was a zero-byte regular file, not the directory shape proper-lockfile creates"), but nobody identified where the regular file came from. Running pi_agent_rust side by side with TS pi is the source.

### Reproduction

```bash
# 1. rust pi has run at least once → lock file exists
ls -l ~/.pi/agent/auth.json.lock   # 0-byte regular file (also settings.json.lock, sessions/session-index.lock)

# 2. any TS pi credential read now fails
node - <<'EOF'
const lockfile = require("proper-lockfile");            // 4.1.2, as vendored by @earendil-works/pi-coding-agent
try { lockfile.lockSync(process.env.HOME + "/.pi/agent/auth.json", { realpath: false }); }
catch (e) { console.log(e.code, e.message); }            // fresh: ELOCKED / stale: ENOTDIR ... rmdir
EOF
```

### Relevant code

- `src/auth.rs:3196-3222` — `auth_lock_path()` / `open_auth_lock_file()`: creates `<auth.json>.lock` with `create(true).truncate(false)`, never removed
- `src/auth.rs:4803-4870` — `lock_file` / `lock_file_shared` / `LockedFile::drop` (fs4 `FileExt`, unlock only)
- TS side: `packages/coding-agent/src/core/auth-storage.ts` (`proper-lockfile` with `lockSync`/`lock`), `proper-lockfile@4.1.2` `lib/lockfile.js` (`mkdir` acquire, `rmdir` stale reclaim)

## Incompatibility 2: OAuth credential serialization

`AuthCredential` is serialized with serde's snake_case variant tags (`src/auth.rs:233-249`):

```rust
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AuthCredential {
    ApiKey { key: String },                              // → "api_key"  (happens to match TS)
    OAuth { access_token, refresh_token, expires, .. },  // → "o_auth"   (TS expects "oauth")
    // ...
}
```

So an OAuth login performed in pi_agent_rust is written as:

```json
{ "type": "o_auth", "access_token": "...", "refresh_token": "...", "expires": 1783598954702 }
```

while TS pi's `OAuthCredentials` shape (`pi-ai` `dist/utils/oauth/types.d.ts`) is:

```json
{ "type": "oauth", "access": "...", "refresh": "...", "expires": 1783598954702 }
```

Both the **tag** (`o_auth` vs `oauth`) and the **field names** (`access_token`/`refresh_token` vs `access`/`refresh`) differ. TS pi compares `cred?.type === "oauth"` exactly and its providers read `credentials.access`, so a Rust-written OAuth credential — even a freshly refreshed, unexpired one — resolves to `No API key for provider: openai-codex`.

The reverse direction is equally broken: a TS-written `{"type": "oauth", "access": ..., "refresh": ...}` entry matches none of the tagged variants and falls into `AuthCredential::Unknown` (`src/auth.rs:283-284`), which maps to `CredentialStatus::Missing` (`src/auth.rs:516`) and `None` during resolution (`src/auth.rs:1094`). **Neither implementation can use an OAuth login performed by the other, despite sharing the same file.** Only `api_key` entries survive, because serde's `api_key` happens to equal the TS tag.

Observed failure chain in practice (TS pi driven via SDK, with a valid unexpired `openai-codex` token on disk that pi_agent_rust had written in its own shape):

```
"stopReason": "error", "errorMessage": "No API key for provider: openai-codex"
```

## Secondary issue: OAuth refresh runs outside the lock

Even with both formats aligned, `refresh_expired_oauth_tokens_with_client()` (`src/auth.rs:660-783`) performs the HTTP refresh **without holding the lock** — the lock is only taken for the final save. Providers with rotating (single-use) refresh tokens race when a rust pi and a TS pi refresh near-simultaneously: the loser's refresh token is already consumed → `invalid_grant` → credentials genuinely invalidated, forcing a re-login. TS pi avoids this within its own ecosystem via lock-then-double-check-then-refresh; matching that would make cross-implementation coexistence safe.

## Suggested fix

1. Adopt a proper-lockfile-compatible protocol for files under the shared `~/.pi/agent/`: acquire = `mkdir("<file>.lock")`, hold = touch mtime periodically, release = `rmdir`, stale = mtime threshold. That makes the two implementations mutually exclusive for real, and eliminates the leftover regular file that poisons the TS side.
2. Serialize OAuth credentials in the upstream shape (`{"type": "oauth", "access": ..., "refresh": ..., "expires": ...}`) and accept that shape on read — serde `rename`/`alias` attributes would keep previously-written `o_auth` entries loadable — instead of routing upstream-shaped entries to `Unknown`.
3. Move the OAuth refresh inside the lock with an expiry double-check after acquiring (as TS pi does).
4. (Mitigation while 1 is pending) at minimum, delete `auth.json.lock` on unlock, and if a *directory* is found at the lock path, wait on it rather than failing with `EISDIR`.

Happy to provide more logs or test the fix. Thanks for the great port!

**Comment by Dicklesworthstone** [comment]
Fixed in ad719ad3 — pi_agent_rust now speaks proper-lockfile's directory-based lock protocol (mkdir-to-acquire / rmdir-to-release, no leftover 0-byte regular file that poisoned TS pi with ENOTDIR) across the auth/settings/session-index lock paths, and (de)serializes OAuth credentials in upstream TS pi's shape so an OAuth login by either implementation is readable by the other. Thanks for the exceptionally thorough writeup!
